docs/HARDWARE-MODEL.md

```diff
@@ -0,0 +1,154 @@
+
+## Testing
+
+The model file as well as other hardware functions used by EVE-OS should be tested before submitting a pull request to add the model file.
+
+### Testing the model file
+
+The testing includes to assign each assignable adapter to an app instance and verifying that it works when used by that app instance. That testing is different if this is a USB controller, audio device, serial port, GPU, or network port.
+While testing this it is also critical to test that other PCI functions on the same controller remain usable by EVE-OS (or that they can be assigned to other application instances and work there). For instance, if eth4, eth5, eth6, and eth7 are separate functions on the same PCI controller. For example,
+
+```shell
+# lspci | grep 0b:00
+0b:00.0 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
+0b:00.1 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
+0b:00.2 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
+0b:00.3 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
+# find /sys/kernel/iommu_groups/ -type l | grep 0b:00
+/sys/kernel/iommu_groups/41/devices/0000:0b:00.3
+/sys/kernel/iommu_groups/38/devices/0000:0b:00.0
+/sys/kernel/iommu_groups/40/devices/0000:0b:00.2
+/sys/kernel/iommu_groups/39/devices/0000:0b:00.1
+```
+
+In such a case it would make sense to assign eth4, eth5, eth6, and eth7 to different application instances and verify that they are able to send and receive packets.
+
+### Testing hardware watchdog
+
+On a test system, ideally connected to a monitor and keyboard, suspend the software watchdog process and wait for some time (based on the hardware watchdog timer, which might be configurable in the BIOS) and the hardware watchdog should reset the box. One way to do this is:
+
+```shell
+# pkill -SUSP watchdog
+```
+
+### Verifying the TPM support
+
+The existence of the TPM is determined by the spec.sh script and it puts this in the hsm attribute. Since EVE only supports TPM 2.0 this does not detect TPM 1.2 devices.
+
+However, even a TPM 2.0 device might not support all of the functionality needed for ECC certificates, ECDH generation, sealing a vault key under the TPM measurements, measured boot and remote attestation.
+
+Once EVE-OS is installed on a device and the device is onboarded to the controller one can check that this functionality is seen by the controller.
+
+Prior to that one can verify that:
+
+- There is a /config/device.cert.pem and no /config/device.key.pem, indicating that the device certificate was generated by the TPM (and the private key is kept in the TPM chip)
+- There is a set of certificate files in /persist/certs/: ek.cert.pem, attest.cert.pem, and ecdh.cert.pem
+
+### Verifying the VT-x and VT-d
+
+This is reported in the EVE API and can be seen in the controller.
```

Comment:
Still it maybe useful to mention to look at /dev/{kvm,xen}

Reply:
Seems like the KVM check for io virtualization checks if /sys/kernel/iommu_groups exists and is non-empty. Is that what makes sense documenting?

Reply:
Yup!